doc/neps/nep-0021-advanced-indexing.rst

```diff
@@ -205,10 +313,13 @@ provide ``__getitem__`` or ``__setitem__``) the special attributes should
 just work. Subclasses that *do* provide it must be updated accordingly
 and should preferably not subclass working versions of these attributes.
 
-All subclasses will inherit the attributes, however, it seems possible
-to test ``subclass.__getitem__.__classobj__`` when getting i.e.
-``subclass.vindex``. If this is not ``ndarray``, the subclass has special
-handling and an ``AttributeError`` can be given.
+All subclasses will inherit the attributes, however, the implementation
+of ``__getitem__`` on these attributes should test
+``subclass.__getitem__ is ndarray.__getitem__``. If not, the
+subclass has special handling for indexing and ``NotImplementedError``
+should be raised, requiring that the indexing attributes is also explicitly
+overwritten. Likewise, implementations of ``__setitem__`` should check to see
```

Comment:
Ok, here's where my comment about constructing special indexing objects should have gone. Just translate `arr.oindex[ind]` to `arr[np.core.oindex_object(ind)]`. As long as `oindex_object` does not attempt to duck-type `tuple`, then the subclass will just crash if it tried to actually inspect the index. For `ma.array` and `memmap`, the index is just passed through, so this would just work.

Reply:
Sounds like an ok suggestion for the implementation. Though such an object would have to explicitly not provide the sequence API or someone might still end up just using it... I doubt it is enough for MA, but it might make the MA implementation simple. Also this can provide help to the duck array, since this "indexing object" can provide extra information like "a copy should be made" or "here I converted the boolean arrays for you".

Reply:
Yep, a class like this would work:

```
class oindex_object:
    value: tuple
```

I think it would work just fine for `ma` - `ma.array.__getitem__` does nothing with `idx` other than `data[idx]` and `mask[idx]`